test/Contract/Buy.purs

```diff
@@ -3,79 +3,217 @@ module Test.Contract.Buy (suite) where
 import Seabug.Contract.Buy (marketplaceBuy')
-import Seabug.Contract.Util (ReturnBehaviour(..))
+import Seabug.Contract.Util (ReturnBehaviour(..), SeabugTxData)
 import Seabug.MarketPlace (marketplaceValidatorAddr)
+import Seabug.Types (MintParams)
 import Test.Contract.Util
-  ( assertContract
+  ( ContractWrapAssertion
+  , assertContract
   , assertLovelaceDecAtAddr
   , assertLovelaceIncAtAddr
   , callMintCnft
   , callMintSgNft
   , checkNftAtAddress
-  , findUtxoWithNft
   , mintParams1
+  , mintParams2
+  , mintParams3
+  , mintParams4
   , plutipConfig
-  , privateStakeKey
+  , privateStakeKey1
+  , privateStakeKey2
   , withAssertions
   )
 import TestM (TestPlanM)
 
+type BuyTestData =
+  { sellerPayAddr :: Address -- The enterprise address of the seller
+  , buyerAddr :: Address -- The address used by the buyer
+  , authorPayAddr :: Address -- The enterprise address of the author
+  , mpScriptAddr :: Address -- The address of the marketplace script
+  , mintParams :: MintParams -- The params used to mint the bought nft
+  }
+
+type PostBuyTestData =
+  { buyTestData :: BuyTestData
+  , txData :: SeabugTxData -- The data of the buy transaction
+  }
+
+type ExpectedShares =
+  { minMpGain :: BigInt
+  , minSellerGain :: BigInt
+  , minAuthorGain :: BigInt
+  }
+
+type BuyTestConfig =
+  { mintParams :: MintParams
+  , expectedShares :: ExpectedShares
+  , retBehaviour :: ReturnBehaviour
+  , postBuyAssertions :: PostBuyTestData -> Array (Contract () Unit)
+  }
+
+buyTestConfig1 :: BuyTestConfig
+buyTestConfig1 =
+  { mintParams: mintParams1
+  , expectedShares:
+      { minMpGain: BigInt.fromInt $ 10 * 1000000
+      , minSellerGain: BigInt.fromInt $ 90 * 1000000
+      , minAuthorGain: BigInt.fromInt $ 90 * 1000000
+      }
+  , retBehaviour: ToMarketPlace
+  , postBuyAssertions: nftToMarketPlaceAssert
+  }
+
+buyTestConfig2 :: BuyTestConfig
+buyTestConfig2 =
+  { mintParams: mintParams2
+  , expectedShares:
+      { minMpGain: BigInt.fromInt 0
+      , minSellerGain: BigInt.fromInt $ 100 * 1000000
+      , minAuthorGain: BigInt.fromInt $ 100 * 1000000
+      }
+  , retBehaviour: ToMarketPlace
+  , postBuyAssertions: nftToMarketPlaceAssert
+  }
+
+buyTestConfig3 :: BuyTestConfig
+buyTestConfig3 =
+  { mintParams: mintParams3
+  , expectedShares:
+      { minMpGain: BigInt.fromInt $ 10 * 1000000
+      , minSellerGain: BigInt.fromInt $ 90 * 1000000
+      , minAuthorGain: BigInt.fromInt $ 90 * 1000000
+      }
+  , retBehaviour: ToMarketPlace
+  , postBuyAssertions: nftToMarketPlaceAssert
+  }
+
+buyTestConfig4 :: BuyTestConfig
+buyTestConfig4 = buyTestConfig2 { mintParams = mintParams4 }
+
 suite :: TestPlanM Unit
 suite =
-  only $ test "Buy" do
-    let
-      distribution =
-        ( withStakeKey privateStakeKey
+  only $ group "Buy" do
+    test "Seller is author, no low prices, nft to marketplace" $
+      mkBuyTest buyTestConfig1
+    skip $ test "Seller is author, no low prices, nft to buyer" $
```

Comment:
Failing until we figure out the issue with the buyer not paying the full price

Reply:
We've decided that this is acceptable, the buyer will pay the min-ada less than the price given.
So lets modify the test here to account for that

Reply:
See the top comment for how we're testing this